man/man1/mbm_multidatalist.1
.TH mbm_multidatalist 1 "18 August 2011" "MB-System 5.0" "MB-System 5.0"
.SH NAME
\fBmbm_multidatalist\fP - Macro to use mbdatalist with the \fB-O\fP option
to generate standard  ancilliary files, using parallel processes.

.SH VERSION
Version 5.0

.SH SYNOPSIS
\fBmbm_multidatalist\fP \fB-I\fP\fIdatalist\fP 
[\fB-X\fP\fInprocesses\fP [\fB-N\fP \fB-H\fP \fB-V\fP]

.SH DESCRIPTION

\fBmbm_multidatalist\fP is a macro used to efficiently generate ancilliary files
for the swath data files referenced in the input datalist file. This is 
accomplished by executing \fBmbdatalist\fP with the \fB-O\fP or \fB-N\fP option 
for each swath file.

Programs such as \fBmbgrid\fP try to check statistics or "inf" 
files to see if the corresponding data files include data within
the specified geographic bounds. Other programs look for
"fast bathymetry" or "fast navigation" ("fbt" or "fnv") files
in order to read the data more quickly. The \fB-N\fP option causes 
\fBmbm_multidatalist\fP to create these three types of ancillary files 
for each swath data file. By default, \fBmbm_multidatalist\fP 
creates the "inf", "fbt", and "fnv" files only when they don't 
already exist or are out of date (older than the data file).
The \fB-N\fP option causes \fBmbm_multidatalist\fP to create 
(or recreate) the ancillary files 
for each swath data file. 

The \fB-X\fP\fInprocesses\fP option sets the number of parallel processes that can be
started by \fBmbm_multidatalist\fP. In general, the most efficient (fastest) execution of 
the overall command will be achieved when \fInprocesses\fP equals the number of 
available CPUs or cores.

.SH AUTHORSHIP
Christian Ferreira
.br
  MARUM, University of Bremen
.br
David W. Caress (caress@mbari.org)
.br
  Monterey Bay Aquarium Research Institute
.br
Dale N. Chayes (dale@ldeo.columbia.edu)
.br
  Lamont-Doherty Earth Observatory

.SH SIMPLE DESCRIPTION OF BASIC OPTIONS
.TP
.B \-H
This "help" flag cause the program to print out a description
of its operation and then exit immediately.
.TP
.B \-C
.br
This option causes the program to check if each output file
already exists, and to forgo copying those that do exist.
.TP
.B \-I
\fIdatalist\fP
.br
Sets the input datalist filename. A datalist is an ascii 
file containing a list of the input swath sonar
data files to be copied and their formats.  
In the \fIdatalist\fP file, each
swath data file should be followed by a data format identifier, e.g.:
 	datafile1 11
 	datafile2 24
.br
As discussed in the \fBMB-System\fP manual page, datalists may reference
other datalists, allowing users to construct recursive datalist structures.
This program uses the \fBMBIO\fP library and will read or write any swath sonar
format supported by \fBMBIO\fP. A list of the swath sonar data formats
currently supported by \fBMBIO\fP and their identifier values
is given in the \fBMBIO\fP manual page.
.TP
.B \-N
This argument causes \fBmbm_multidatalist\fP to generate the three ancillary
data files ("inf", "fbt", and "fnv") even if they already exist and are up to date. 
.TP
.B \-V
Causes \fBmbm_multidatalist\fP to operate in "verbose" mode 
so that it outputs
more information than usual.
.TP
.B \-X
\fInprocesses\fP
.br
This option sets the number of parallel, simultaneous \fBmbdatalist\fP processes that may be
run.

.SH EXAMPLES
Suppose one has two format 88 Reson 7125 data files:
 	20110805_053612.mb88
 	20110805_054149.mb88
.br
One can generate a datalist called datalistr.mb-1 referencing these
three swath files using a text editor or by using \fBawk\fP:
 	ls -1 *.mb88 | awk '{print $1" 88"}' > datalist.mb-1
.br
In order to generate the usual ancilliary files (*.inf, *.fbt, and *.fnv files)
use the \fBmbm_multidatalist\fP macro:
 	mbm_multidatalist -I datalist.mb-1 -X2 -V
.br
The macro prints out the following as it generates the *.inf, *.fbt, and *.fnv
files for each swath data file:

 	Running mbm_multidatalist...
 	Executing mbdatalist -I20110805_053612.mb88 -O -V
 	Executing mbdatalist -I20110805_054149.mb88 -O -V
 	
 	Program mbdatalist
 	
 	Version $Id$
 	Program mbdatalist
 	MB-system Version 5.3.1904
 	Version $Id$
 	MB-system Version 5.3.1904
 	
 	Generating inf file for 20110805_054149.mb88
 	
 	Generating inf file for 20110805_053612.mb88
 	Generating fbt file for 20110805_054149.mb88
 	Generating fbt file for 20110805_053612.mb88
 	Generating fnv file for 20110805_054149.mb88
 	Generating fnv file for 20110805_053612.mb88
 	
 	Total swath files:         1
 	
 	Total swath files:         1

Note that the terminal messages from the two \fBmbdatalist\fP processes are
mixed because the processes are executed in parallel.

.SH SEE ALSO
\fBmbsystem\fP(l), \fBmbdatalist\fP(l), \fBmbcopy\fP(l), 
\fBmbdatalist\fP(l)

.SH BUGS
Perhaps. 




branches/autobuild-trunk/src/macros/mbm_multidatalist.pl
eval '(exit $?0)' && eval 'exec perl -S $0 ${1+"$@"}'
                    & eval 'exec perl -S $0 $argv:q'
                         if 0;
#--------------------------------------------------------------------
#    The MB-system:	mbm_multidatalist.perl	7/18/2011
#   $Id: mbm_multidatalisz.pl 19XX 2011-07-18 15:40:00Z ferreira $
#
#    Copyright (c) 2011 by 
#    D. W. Caress (caress@mbari.org)
#      Monterey Bay Aquarium Research Institute
#      Moss Landing, CA
#    and D. N. Chayes (dale@lamont.ldgo.columbia.edu)
#      Lamont-Doherty Earth Observatory
#      Palisades, NY  10964
#
#    See README file for copying and redistribution conditions.
#--------------------------------------------------------------------
#
# Command:
#   mbm_multidatalist
#
# Purpose:
#   Perl shellscript to generate standard ancilliary files using 
#   parallel mbdatalist -O processes
#
# Usage:
#   mbm_multidatalist -Xcpu -Idatalist
#
# Author:
#   Christian Ferreira
#   MARUM - Center for Marine Environmental Sciences
#   Klagenfurter Str. 2, GEO Building
#   28359, Bremen, Germany
#   July 17, 2011
#
# Version:
#   $Id: mbm_multidatalist.pl 19XX 2011-07-18 15:40:00Z ferreira $
#
# Revisions:
#
#  Revision 1.0		2011-07-18 15:40:00	cferreira
#  Initial version with partial implementation of mbdatalist flags
#

$program_name = "mbm_multidatalist";

# Deal with command line arguments
$command_line = "@ARGV";
&Getopts('I:i:NnX:x:HhVv');
$help =    		($opt_H || $opt_h);
$datalist =    		($opt_I || $opt_i || "datalist.mb-1");
$regenerate =    	($opt_N || $opt_n);
$verbose = 		($opt_V || $opt_v);
$cpu =        		($opt_X || $opt_x || 1);

# print out help message if required
if ($help)
	{
	print "\n$program_name:\n";
	print "\nVersion: mbm_multidatalist.pl 19XX 2011-07-18 15:40:00Z ferreira\n";
	print "\nMacro to use mbdatalist with the -O option to generate standard ancilliary files, using parallel processes\n";
	print "\n";
	print "\nBasic Usage: \n\n";
	print "\t$program_name -Idatalist -Xnprocesses [-N -H -V]\n\n";
	exit 0;
	}

# tell the world we got started
if ($verbose) 
	{
	print "\nRunning $program_name...\n";
	}

# exiting if CPU number is lower than 1
if ($cpu < 1) 
	{
	print "\nExiting $program_name, CPU(s) number(s) should be\n equal or bigger than one...\n\n";
	exit 0;
	}
	
# set the mbdatalist control option
if ($regenerate)
	{
	$option = "-N";
	}
else
	{
	$option = "-O";
	}

# loop over all files referenced in datalist
@mbdatalist = `mbdatalist -F-1 -I$datalist`;
while (@mbdatalist)
	{
	$line = shift @mbdatalist;
	if ($line =~ /(\S+)\s+(\S+)/)
		{
		($file_mb,$format_mb) = 
			$line =~ /(\S+)\s+(\S+)/;
		push(@files_data, $file_mb);
		push(@formats, $format_mb);
		}
	}

# start parellel computing
use Parallel::ForkManager;
use LWP::Simple;
my $pm=new Parallel::ForkManager($cpu);
  
foreach $file_mb (@files_data)
	{
	my $pid = $pm->start and next;
	if ($verbose) 
		{
		$command = "mbdatalist -I$file_mb $option -V";
		print "Executing $command\n";
		}
	else
		{
		$command = "mbdatalist -I$file_mb $option";
		}
	system $command;
	$pm->finish;
	}

$pm->wait_all_children;

exit 0;

#-----------------------------------------------------------------------
# This should be loaded from the library but its safer to
# just include it....
#
;# getopts.pl - a better getopt.pl

;# Usage:
;#      do Getopts('a:bc');  # -a takes arg. -b & -c not. Sets opt_* as a
;#                           #  side effect.

sub Getopts {
    local($argumentative) = @_;
    local(@args,$_,$first,$rest);
    local($errs) = 0;
    local($[) = 0;

    @args = split( / */, $argumentative );
    while(@ARGV && ($_ = $ARGV[0]) =~ /^-(.)(.*)/) {
     ($first,$rest) = ($1,$2);
     $pos = index($argumentative,$first);
     if($pos >= $[) {
         if($args[$pos+1] eq ':') {
          shift(@ARGV);
          if($rest eq '') {
              ++$errs unless @ARGV;
              $rest = shift(@ARGV);
          }
          eval "\$opt_$first = \$rest;";
         }
         else {
          eval "\$opt_$first = 1";
          if($rest eq '') {
              shift(@ARGV);
          }
          else {
              $ARGV[0] = "-$rest";
          }
         }
     }
     else {
         print STDERR "Unknown option: $first\n";
         ++$errs;
         if($rest ne '') {
          $ARGV[0] = "-$rest";
         }
         else {
          shift(@ARGV);
         }
     }
    }
    $errs == 0;
}
